### ISSUE TITLE
managing different i18n domains

### ISSUE THREAD

**Issue by faassen** [yahoo/react-intl#111]
I've been doing some thinking on how to manage i18n domains. Here are some use cases:

* have multiple independently developed (React) applications that can be composed into a larger application.

* messages have a message id, i.e the key in the messages mapping.

* each application has its own messages mapping in a "domain". This so you can have short, scoped message ids and applications can be isolated from each other.

* domain data is loaded from the server, or taken from client-side data; this can be configured.

* formats are shared globally, though it may also be domain specific. I haven't thought much about this yet and will ignore it for now and focus on messages.

Does anyone have similar use cases?

Here's a sketch of how this might work:

* Domains have a domain id, so that things can refer to a domain.

* There's a global message id database (or if not global, a top-level context) that has a mapping from domain id to messages.

* Application code can ensure that a domain is in the database before rendering starts, it can do this by asking the database to load the domain. To this end, the database needs to be constructed either with all needed data or with a loading strategy for getting a domain database loaded from the server or wherever else.

* When you use a message id in an application, it looks up the message in the application's domain first.

In React terms:

`IntlDatabase` component sitting at the top of the rendering tree, given a loadDomain function prop optionally, and domains prop (can give both, in which case the domains props contains preloaded domains). This has an API that is used to get the messages for a domain id. It sets up a context which is used by...

`IntlDomain` component. This gets an prop `domain` which contains the domain id to use in the underlying components. Uses the API on `IntlDatabase` to get messages for a domain. Offers the `getIntlMessage` API. This sets up a context which is used by...

`FormatMessage` is adjusted so it gets a `messageId` prop which looks up the message in the domain context.

What do people think about this design? Anything I overlooked?

**Comment by ericf** [comment]
I don't think adding new components around the concept of domains is a direction that we should take this library. Instead, I would just like to make sure that someone _could_ implement this behavior in their app using React Intl.

In the current version of the library you should be able to pass the `messages` and `formats` into independent apps, and delegate the responsibility to the app's owner/composer. And/or not use `getIntlMessage()`, but rather use you're own utility function to find messages based on your domain-model implementation.

**Comment by faassen** [comment]
Thanks for getting back to this! I was not suggesting that all of this should be added to the core library, but that the core library should support these patterns so that they can be implemented by another library.

To get more concrete, I built such a library. Since I had to actually make stuff work it turned out quite different from my sketch. The main reason is that I ran into huge problems managing contexts with React so I went another way. I am still holding out hope for contexts, but I think that will have to wait for a new version of React with more sane context handling.

The library is here:

https://www.npmjs.com/package/react-intl-db

Hopefully it can help inform this discussion.

The basic idea is that there's a database that manages messages in domains. You then can create a shortcut `<Format>` component that gets the message id for a domain.

Another goal of this library was to avoid having to explicitly use the mixin; I use ES6 classes in my React code and I don't want to switch away from this.

How could the core library help? Quite independently from my goals here, I dislike having to use a utility function to get the message for a message id each time; the framework should be able to do this for me. It would be nice if the core `FormatMessage` had a `messageId` prop. By default it'd just look in the context for the message, but there should be a way to plug in a policy that determines how the actual message is looked up.

Once nicer context tools land in React, another external library can then offer a component to set up the domain in the context; this can then be used by a custom messageId lookup policy that it offers.

There is another use case though: it is nice to have all messages for all domains loaded before the application starts rendering. In `react-intl-db` I determine this by knowing what custom `Format` components have been created. If the domains are embedded in the React component trees, you can't really do this. Waiting with rendering lower down in the tree until messages for its domain have been loaded is somewhat awkward (I do this only in the top level application component now), though does have the benefit of working with run-time composed applications. This whole use case needs further thought.

**Comment by faassen** [comment]
Yes, I saw that issue; I actually commented on it too already. If there is a way to override how message ids are looked up then it becomes feasible to implement domain-based lookup in an extension.

**Comment by ericf** [comment]
Closing this for now, since I'm not sure there's anything actionable right now.
